### PR TITLE
 fix disposing issues and incorrect work in netcore2.0

### PR DESCRIPTION
**Pull request by marfenij**
So, I extend test projects to several target frameworks, remove unsupported stuff like RhinoMocks, SharpTestsEx; append test to check close/dispose workflow, and check behavior when buffer sizes changes (at this moment I got exceptions on real project)

### CODE REVIEW

File: test/protobuf-net-data.Tests/protobuf-net-data.Tests.csproj [hunk 1]
```diff
@@ -3,21 +3,48 @@
   <PropertyGroup>
     <DebugType Condition="$(codecov) == 'true'">full</DebugType>
     <IsPackable>false</IsPackable>
-    <TargetFramework>net471</TargetFramework>
+    <TargetFrameworks>net40;net47;net471;netcoreapp2.0</TargetFrameworks>
```
Comment (dotarj) [inline test/protobuf-net-data.Tests/protobuf-net-data.Tests.csproj]: Target framework net47 could be removed, right?

File: .gitignore [hunk 2]
```diff
@@ -98,6 +98,13 @@ ipch/
 *.cachefile
 *.VC.db
 *.VC.VC.opendb
+obj/
+[Bb]in
+[Dd]ebug*/
+[Rr]elease*/
+Ankh.NoLoad
+.nuget/nuget.exe
+.vs/
```
Comment (dotarj) [inline .gitignore]: .vs/ is already mentioned on line 40.

File: .gitignore [hunk 3]
```diff
@@ -98,6 +98,13 @@ ipch/
 *.cachefile
 *.VC.db
 *.VC.VC.opendb
+obj/
```
Comment (dotarj) [inline .gitignore]: obj/ is already mentioned on line 36.

File: .gitignore [hunk 4]
```diff
@@ -98,6 +98,13 @@ ipch/
...
+[Bb]in
```
Comment (dotarj) [inline .gitignore]: Already mentioned on line 28 - 37.

File: .gitignore [hunk 5]
```diff
@@ -98,6 +98,13 @@ ipch/
...
+[Dd]ebug*/
+[Rr]elease*/
+Ankh.NoLoad
```
Comment (dotarj) [inline .gitignore]: Not sure what this is (Ankh.NoLoad).

File: .gitignore [hunk 6]
```diff
@@ -98,6 +98,13 @@ ipch/
 *.cachefile
```
Comment (dotarj) [inline .gitignore]: The changes made to .gitignore mostly seem to be double entries, better drop the changes to .gitignore.

File: src/protobuf-net-data/ProtoDataReader.cs [hunk 7]
```diff
@@ -341,8 +341,7 @@ public bool IsDBNull(int i)
 
         public void Close()
         {
-            stream.Close();
-            IsClosed = true;
+            this.Dispose();
```
[the remark this answers is not in the document]
Comment (marfenij) [reply src/protobuf-net-data/ProtoDataReader.cs]: You are right - correct way call Close() in Dispose method

File: test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetBytesMethod.cs [hunk 8]
```diff
@@ -88,7 +88,7 @@ public void ShouldThrowExceptionWhenBufferOffsetIsLessThanZero()
                 dataReader.Read();
 
                 // Assert
-                Assert.Throws<ArgumentOutOfRangeException>("bufferOffset", () => dataReader.GetBytes(0, 0, new byte[1], -1, 1));
```
[the remark this answers is not in the document]
Comment (marfenij) [reply test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetBytesMethod.cs]: I understand, but old xunit, which used for net40 not support that argument

File: test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetValuesMethod.cs [hunk 9]
```diff
@@ -59,7 +59,7 @@ public void ShouldReturnCorrespondingValues()
                 dataReader.GetValues(result);
 
                 // Assert
-                Assert.Equal(new[] { value }, result);
+                Assert.Equal(new object[] { value }, result);
```
[the remark this answers is not in the document]
Comment (marfenij) [reply test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetValuesMethod.cs]: I know, but I really  remember what only explicitly using type do not cause compile error

File: test/protobuf-net-data.Tests/protobuf-net-data.Tests.csproj [hunk 10]
```diff
@@ -3,21 +3,48 @@
...
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <IncludeTestSdk>true</IncludeTestSdk>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(TargetFramework)' == 'net20'">
```
[the remark this answers is not in the document]
Comment (marfenij) [reply test/protobuf-net-data.Tests/protobuf-net-data.Tests.csproj]: Yeah, I'd try make test on net20, but it's pain )
I clean up that

File: test/protobuf-net-data.Tests/protobuf-net-data.Tests.csproj [hunk 11]
```diff
@@ -3,21 +3,48 @@
...
-    <PackageReference Include="xunit" Version="2.3.1" />
+  </ItemGroup>
+
+  <ItemGroup>
     <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
-    <PackageReference Include="RhinoMocks" Version="3.6.1" />
-    <PackageReference Include="SharpTestsEx" Version="2.0.0" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp2.0' or '$(TargetFramework)' == 'netstandard2.0'">
```
[the remark this answers is not in the document]
Comment (marfenij) [reply test/protobuf-net-data.Tests/protobuf-net-data.Tests.csproj]: yes, test project should not use netstandard as target framework

File: test/protobuf-net-data.Tests/protobuf-net-data.Tests.csproj (hunk 1 above)
Comment (marfenij) [reply test/protobuf-net-data.Tests/protobuf-net-data.Tests.csproj]: it's possible ))

File: .gitignore (hunk 6 above)
Comment (marfenij) [reply .gitignore]: sorry for that, i revert .gitignore to original state https://www.gitignore.io/api/visualstudio,visualstudiocode

File: test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetBytesMethod.cs [hunk 12]
```diff
@@ -100,7 +100,7 @@ public void ShouldThrowExceptionWhenBufferOffsetIsGreaterThanBufferSize()
                 dataReader.Read();
 
                 // Assert
-                Assert.Throws<ArgumentOutOfRangeException>("bufferOffset", () => dataReader.GetBytes(0, 0, new byte[1], 10, 1));
```
[the remark this answers is not in the document]
Comment (marfenij) [reply test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetBytesMethod.cs]: As I mentioned above, this is old xUnit issue

File: test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetBytesMethod.cs [hunk 13]
```diff
@@ -112,7 +112,7 @@ public void ShouldThrowExceptionWhenBufferOffsetIsEqualToBufferSize()
                 dataReader.Read();
 
                 // Assert
-                Assert.Throws<ArgumentOutOfRangeException>("bufferOffset", () => dataReader.GetBytes(0, 0, new byte[1], 1, 1));
```
[the remark this answers is not in the document]
Comment (marfenij) [reply test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetBytesMethod.cs]: As I mentioned above, this is old xUnit issue

File: test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetCharsMethod.cs [hunk 14]
```diff
@@ -88,7 +88,7 @@ public void ShouldThrowExceptionWhenBufferOffsetIsLessThanZero()
                 dataReader.Read();
 
                 // Assert
-                Assert.Throws<ArgumentOutOfRangeException>("bufferOffset", () => dataReader.GetChars(0, 0, new char[1], -1, 1));
```
[the remark this answers is not in the document]
Comment (marfenij) [reply test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetCharsMethod.cs]: As I mentioned above, this is old xUnit issue

File: test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetCharsMethod.cs [hunk 15]
```diff
@@ -100,7 +100,7 @@ public void ShouldThrowExceptionWhenBufferOffsetIsGreaterThanBufferSize()
                 dataReader.Read();
 
                 // Assert
-                Assert.Throws<ArgumentOutOfRangeException>("bufferOffset", () => dataReader.GetChars(0, 0, new char[1], 10, 1));
```
[the remark this answers is not in the document]
Comment (marfenij) [reply test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetCharsMethod.cs]: As I mentioned above, this is old xUnit issue

File: test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetCharsMethod.cs [hunk 16]
```diff
@@ -112,7 +112,7 @@ public void ShouldThrowExceptionWhenBufferOffsetIsEqualToBufferSize()
                 dataReader.Read();
 
                 // Assert
-                Assert.Throws<ArgumentOutOfRangeException>("bufferOffset", () => dataReader.GetChars(0, 0, new char[1], 1, 1));
```
[the remark this answers is not in the document]
Comment (marfenij) [reply test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetCharsMethod.cs]: As I mentioned above, this is old xUnit issue

File: test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetValuesMethod.cs [hunk 17]
```diff
@@ -19,7 +19,7 @@ public void ShouldThrowExceptionWhenValuesIsNull()
                 dataReader.Close();
 
                 // Assert
-                Assert.Throws<ArgumentNullException>("values", () => dataReader.GetValues(null));
```
[the remark this answers is not in the document]
Comment (marfenij) [reply test/protobuf-net-data.Tests/ProtoBufDataReaderTests.TheGetValuesMethod.cs]: As I mentioned above, this is old xUnit issue

**Comment by marfenij** [comment]
Thank you for review. I try made small changes for PR in future )) (also I know that feature I missed include, and that feature helps me debug that bug with buffers sizes)

So, on of problem it's incorrect assuming that protobuf-net always return 1024 byte in buffer.
And another thing reproduce only in netcore20, this guys made optimization for allocation memory for buffers - now buffer size equal to some power of 2 (stream copy try use standard buffer size 80k, but in netcore20 allocate buffer in 128k)

this two thing cause runtime exception that internal (circular) buffer overflow

**Comment by marfenij** [comment]
@dotarj also some trouble with MSBuild.SonarQube.Runner on AppVeyor build

**Comment by dotarj** [comment]
Thanks for updating the PR, it's merged.

There is a (sonarqube) problem with the build with pull requests from forks (secure variables are not available for security). I'll fix this later.

**Comment by dotarj** [comment]
@marfenij A new version of protobuf-net-data containing your fix is [released](https://github.com/dotarj/protobuf-net-data/releases/tag/2.1.1).

**Comment by marfenij** [comment]
@dotarj great!!

**Pull request merged**